Sum Types: fake type mismatch reported by the C compiler

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
CC version: cc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9)
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg db_test.v`
```v
module test

type Opt<T> = None<T> | Some<T>

struct None<T> {}

struct Some<T> {
mut:
	value T
}

fn operation(r int) Opt<int> {
	return if r > 0 {
		Some<int>{r}
	} else {
		None<int>{}
	}
}

fn test_op(){
	op := operation(1)
	assert Opt<int>(Some<int>{1}) == op
}
```

**What did you expect to see?**

Code should compile


**What did you see instead?**
```
==================
/tmp/v_1000/db_test.3409957231411116828.tmp.c:12599: error: type mismatch in conditional expression (have 'struct test__Some_T_int' and 'struct test__None_T_int')
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/if.v
+++ b/vlib/v/checker/if.v
@@ -204,6 +204,7 @@ pub fn (mut c Checker) if_expr(mut node ast.IfExpr) ast.Type {
 						}
 					}
 					if node.is_expr && c.table.sym(former_expected_type).kind == .sum_type {
+						node.typ = former_expected_type
 						continue
 					}
 					if is_noreturn_callexpr(last_expr.expr) {

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -74,7 +74,12 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 				g.expr(branch.cond)
 				g.write(' ? ')
 			}
+			prev_expected_cast_type := g.expected_cast_type
+			if node.is_expr && g.table.sym(node.typ).kind == .sum_type {
+				g.expected_cast_type = node.typ
+			}
 			g.stmts(branch.stmts)
+			g.expected_cast_type = prev_expected_cast_type
 		}
 		if node.branches.len == 1 {
 			g.write(': 0')
@@ -195,11 +200,12 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 			}
 		}
 		if needs_tmp_var {
+			prev_expected_cast_type := g.expected_cast_type
 			if node.is_expr && g.table.sym(node.typ).kind == .sum_type {
 				g.expected_cast_type = node.typ
 			}
 			g.stmts_with_tmp_var(branch.stmts, tmp)
-			g.expected_cast_type = 0
+			g.expected_cast_type = prev_expected_cast_type
 		} else {
 			// restore if_expr stmt header pos
 			stmt_pos := g.nth_stmt_pos(0)

--- a/vlib/v/tests/if_expr_with_generic_sumtype_test.v
+++ b/vlib/v/tests/if_expr_with_generic_sumtype_test.v
@@ -1,0 +1,17 @@
+type Opt<T> = None<T> | Some<T>
+
+struct None<T> {}
+
+struct Some<T> {
+mut:
+	value T
+}
+
+fn operation(r int) Opt<int> {
+	return if r > 0 { Some<int>{r} } else { None<int>{} }
+}
+
+fn test_if_expr_with_generic_sumtype() {
+	op := operation(1)
+	assert Opt<int>(Some<int>{1}) == op
+}

